File: Прикладное Python/ЛЗ 2/zad_2.py
"""
https://stepik.org/lesson/13377/step/7?unit=3571
"""

"""
Предположим, что в программе определён макрос sqr.

#define sqr(x) x * x


Какое значение будет иметь следующее выражение?

sqr(3 + 0)

"""
# в с++ данная фунция вывела бы 3, но так как мы будем использовать lambda фунцкию то результат будет равен 9


#реализация

func =  lambda x: x*x

print(func(3 + 0))

File: Прикладное программирование/ЛЗ 3/zad 2.cpp
/*https://stepik.org/lesson/538/step/9?unit=861*/

/*В коде программы определена следующая функция:

int foo(int n) {
    if (n <= 0)
        return 1;
    return foo((n * 2) / 3) + foo(n - 2);
}
Нужно посчитать, сколько всего раз будет вызвана функция foo, если ее вызвать с аргументом 3 (т.е. foo(3)).
 Самый первый вызов тоже нужно посчитать.

*/

#include <iostream>

using namespace std;

// функция foo будет вызываться 9 раз, чтобы в этом убедиться сделаем глобальныую переменную cnt со счётчиком и посчитаем в программе:

extern int cnt = 0;

int foo(int n) {
    cnt ++;
    if (n <= 0)
        return 1;
    return foo((n * 2) / 3) + foo(n - 2);

}
int main()
{
    int n = 3;
    cout << foo(n) << endl;
    cout << cnt << endl;

    return 0;
}
